File: agent-support/vscode/src/reporting/reporting-profile-service.ts
import { spawn } from "child_process";
import * as vscode from "vscode";
import { getGitAiBinary, resolveGitAiBinary } from "../utils/binary-path";
import {
  DEFAULT_METRICS_API_BASE_URL,
  emptyReportingProfile,
  mergeReportingSettings,
  normalizeOrganizationOptions,
  normalizeReportingSettings,
  resolveOrganizationOptionsUrl,
  type OrganizationOptions,
  type ReportingProfile,
  type ReportingSettings,
} from "./reporting-profile-utils";

const ORGANIZATION_CACHE_KEY = "git-ai.reporting.organization-options";
const ORGANIZATION_CACHE_MAX_AGE_MS = 7 * 24 * 60 * 60 * 1000;
const ORGANIZATION_REQUEST_TIMEOUT_MS = 8_000;
const CLI_TIMEOUT_MS = 10_000;
const MAX_CLI_OUTPUT_BYTES = 1_000_000;

interface OrganizationCacheEntry {
  fetchedAt: number;
  options: OrganizationOptions;
}

interface OrganizationCache {
  [endpoint: string]: OrganizationCacheEntry;
}

export interface OrganizationLoadResult {
  endpoint?: string;
  options?: OrganizationOptions;
  source?: "server" | "cache";
  fetchedAt?: number;
  error?: string;
}

export interface InitialReportingState {
  settings: ReportingSettings;
  importedFields: string[];
  organization: OrganizationLoadResult;
  cliError?: string;
}

export class ReportingProfileService {
  public constructor(private readonly context: vscode.ExtensionContext) {}

  public async loadInitialState(): Promise<InitialReportingState> {
    const kiloSettings = this.readKiloSettings();
    let savedSettings: ReportingSettings = {
      metricsApiBaseUrl: "",
      profile: emptyReportingProfile(),
    };
    let cliError: string | undefined;
    try {
      savedSettings = await this.readGitAiSettings();
    } catch (error) {
      cliError = toErrorMessage(error, "无法读取 Git AI CLI 配置");
    }

    const settings = mergeReportingSettings(savedSettings, kiloSettings, DEFAULT_METRICS_API_BASE_URL);
    const importedFields = importedMissingFields(savedSettings, kiloSettings);
    return {
      settings,
      importedFields,
      organization: await this.loadOrganizationOptions(settings.metricsApiBaseUrl),
      cliError,
    };
  }

  public async loadOrganizationOptions(rawUrl: string): Promise<OrganizationLoadResult> {
    let endpoint: string;
    try {
      endpoint = resolveOrganizationOptionsUrl(rawUrl);
    } catch (error) {
      return { error: toErrorMessage(error, "上报服务器地址无效") };
    }

    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), ORGANIZATION_REQUEST_TIMEOUT_MS);
      let response: Response;
      try {
        response = await fetch(endpoint, { signal: controller.signal, headers: { Accept: "application/json" } });
      } finally {
        clearTimeout(timeout);
      }
      if (!response.ok) {
        throw new Error(`组织架构服务返回 HTTP ${response.status}`);
      }
      const options = normalizeOrganizationOptions(await response.json());
      const fetchedAt = Date.now();
      const cache = this.context.globalState.get<OrganizationCache>(ORGANIZATION_CACHE_KEY) ?? {};
      cache[endpoint] = { fetchedAt, options };
      await this.context.globalState.update(ORGANIZATION_CACHE_KEY, cache);
      return { endpoint, options, source: "server", fetchedAt };
    } catch (error) {
      const cache = this.context.globalState.get<OrganizationCache>(ORGANIZATION_CACHE_KEY) ?? {};
      const cached = cache[endpoint];
      if (cached && Date.now() - cached.fetchedAt <= ORGANIZATION_CACHE_MAX_AGE_MS) {
        return {
          endpoint,
          options: cached.options,
          source: "cache",
          fetchedAt: cached.fetchedAt,
          error: `${toErrorMessage(error, "组织架构服务不可用")}；正在使用上次成功加载的数据`,
        };
      }
      return { endpoint, error: toErrorMessage(error, "无法加载组织架构") };
    }
  }

  public async save(settings: ReportingSettings): Promise<ReportingSettings> {
    const normalized = normalizeReportingSettings(settings);
    const payload = JSON.stringify({
      metrics_api_base_url: normalized.metricsApiBaseUrl,
      reporting_profile: toCliProfile(normalized.profile),
    });
    await this.runGitAi(["config", "reporting-profile", "set", "--stdin"], payload);
    return this.readGitAiSettings();
  }

  private readKiloSettings(): ReportingSettings {
    const configuration = vscode.workspace.getConfiguration("kilo-code");
    const get = (key: string): string => {
      const value = configuration.get<unknown>(key);
      return typeof value === "string" ? value : "";
    };
    return {
      metricsApiBaseUrl: get("aiCodeStatsWebhookUrl"),
      profile: {
        departmentName: get("aiCodeStatsDepartmentName"),
        officeName: get("aiCodeStatsOfficeName"),
        teamName: get("aiCodeStatsTeamName"),
        userName: get("aiCodeStatsUserName"),
        userEmail: get("aiCodeStatsUserEmail"),
      },
    };
  }

  private async readGitAiSettings(): Promise<ReportingSettings> {
    const output = await this.runGitAi(["config", "reporting-profile"]);
    const raw = JSON.parse(output) as { metrics_api_base_url?: unknown; reporting_profile?: unknown };
    const rawProfile = raw.reporting_profile && typeof raw.reporting_profile === "object"
      ? raw.reporting_profile as Record<string, unknown>
      : {};
    return normalizeReportingSettings({
      metricsApiBaseUrl: typeof raw.metrics_api_base_url === "string" ? raw.metrics_api_base_url : "",
      profile: {
        departmentName: stringField(rawProfile.department_name),
        officeName: stringField(rawProfile.office_name),
        teamName: stringField(rawProfile.team_name),
        userName: stringField(rawProfile.user_name),
        userEmail: stringField(rawProfile.user_email),
      },
    });
  }

  private async runGitAi(args: string[], input?: string): Promise<string> {
    await resolveGitAiBinary();
    return new Promise((resolve, reject) => {
      const child = spawn(getGitAiBinary(), args, {
        windowsHide: true,
        stdio: ["pipe", "pipe", "pipe"],
      });
      let stdout = "";
      let stderr = "";
      let outputBytes = 0;
      const timeout = setTimeout(() => {
        child.kill();
        reject(new Error("Git AI CLI 响应超时，请确认 CLI 已安装并重启 VS Code 后重试"));
      }, CLI_TIMEOUT_MS);
      const collect = (target: "stdout" | "stderr", chunk: Buffer) => {
        outputBytes += chunk.length;
        if (outputBytes > MAX_CLI_OUTPUT_BYTES) {
          child.kill();
          reject(new Error("Git AI CLI 输出过大，已中止本次操作"));
          return;
        }
        if (target === "stdout") {
          stdout += chunk.toString("utf8");
        } else {
          stderr += chunk.toString("utf8");
        }
      };
      child.stdout.on("data", (chunk: Buffer) => collect("stdout", chunk));
      child.stderr.on("data", (chunk: Buffer) => collect("stderr", chunk));
      child.on("error", (error) => {
        clearTimeout(timeout);
        reject(new Error(`无法启动 Git AI CLI：${error.message}`));
      });
      child.on("close", (code) => {
        clearTimeout(timeout);
        if (code === 0) {
          resolve(stdout.trim());
        } else {
          reject(new Error((stderr || stdout || `Git AI CLI 退出码 ${code}`).trim()));
        }
      });
      if (input !== undefined) {
        child.stdin.write(input);
      }
      child.stdin.end();
    });
  }
}

function importedMissingFields(saved: ReportingSettings, kilo: ReportingSettings): string[] {
  const fields: Array<[string, string, string]> = [
    ["上报服务器地址", saved.metricsApiBaseUrl, kilo.metricsApiBaseUrl],
    ["部门", saved.profile.departmentName, kilo.profile.departmentName],
    ["处", saved.profile.officeName, kilo.profile.officeName],
    ["组", saved.profile.teamName, kilo.profile.teamName],
    ["姓名", saved.profile.userName, kilo.profile.userName],
    ["公司邮箱", saved.profile.userEmail, kilo.profile.userEmail],
  ];
  return fields.filter(([, current, imported]) => !current.trim() && Boolean(imported.trim())).map(([label]) => label);
}

function toCliProfile(profile: ReportingProfile): Record<string, string | undefined> {
  return {
    department_name: profile.departmentName,
    office_name: profile.officeName,
    team_name: profile.teamName || undefined,
    user_name: profile.userName,
    user_email: profile.userEmail,
  };
}

function stringField(value: unknown): string {
  return typeof value === "string" ? value : "";
}

function toErrorMessage(error: unknown, fallback: string): string {
  return error instanceof Error && error.message ? error.message : fallback;
}

File: agent-support/intellij/src/main/kotlin/org/jetbrains/plugins/template/reporting/ReportingProfileService.kt
package org.jetbrains.plugins.template.reporting

import org.jetbrains.plugins.template.services.GitAiService
import java.io.ByteArrayOutputStream
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.util.concurrent.CompletableFuture
import java.util.concurrent.TimeUnit

class ReportingProfileService(
    private val gitAiService: GitAiService = GitAiService.getInstance(),
    private val kiloImporter: KiloReportingProfileImporter = KiloReportingProfileImporter(),
    private val cache: OrganizationOptionsCache = OrganizationOptionsCache.getInstance(),
) {
    private val httpClient = HttpClient.newBuilder().connectTimeout(Duration.ofSeconds(8)).build()

    fun loadInitialState(): InitialReportingState {
        val kiloSettings = kiloImporter.read()
        var savedSettings = ReportingSettings()
        var cliError: String? = null
        try {
            savedSettings = readGitAiSettings()
        } catch (error: Exception) {
            cliError = error.message ?: "无法读取 Git AI CLI 配置"
        }
        var settings = ReportingSupport.mergeSettings(savedSettings, kiloSettings)
        val organization = loadOrganizationOptions(settings.metricsApiBaseUrl)
        val selectedOffice = organization.options
            ?.departments?.find { it.name == settings.profile.departmentName }
            ?.offices?.find { it.name == settings.profile.officeName }
        if (selectedOffice != null && selectedOffice.teams.isEmpty()) {
            settings = settings.copy(profile = settings.profile.copy(teamName = ""))
        }
        val importedFields = ReportingSupport.importedMissingFields(savedSettings, kiloSettings)
            .filterNot { it == "组" && selectedOffice?.teams?.isEmpty() == true }
        return InitialReportingState(
            settings = settings,
            importedFields = importedFields,
            organization = organization,
            cliError = cliError,
        )
    }

    fun loadOrganizationOptions(rawUrl: String): OrganizationLoadResult {
        val endpoint = try {
            ReportingSupport.resolveOrganizationOptionsUrl(rawUrl)
        } catch (error: IllegalArgumentException) {
            return OrganizationLoadResult(error = error.message ?: "上报服务器地址无效")
        }
        return try {
            val request = HttpRequest.newBuilder(URI(endpoint))
                .timeout(Duration.ofSeconds(8))
                .header("Accept", "application/json")
                .GET()
                .build()
            val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString(StandardCharsets.UTF_8))
            if (response.statusCode() !in 200..299) throw IllegalStateException("组织架构服务返回 HTTP ${response.statusCode()}")
            val options = ReportingSupport.normalizeOrganizationOptions(response.body())
            val fetchedAt = System.currentTimeMillis()
            cache.put(endpoint, fetchedAt, options)
            OrganizationLoadResult(endpoint, options, OrganizationSource.SERVER, fetchedAt)
        } catch (error: Exception) {
            val cached = cache.get(endpoint)
            val cachedOptions = cached?.takeIf { System.currentTimeMillis() - it.fetchedAt <= ORGANIZATION_CACHE_MAX_AGE_MS }
                ?.let(cache::readOptions)
            if (cached != null && cachedOptions != null) {
                OrganizationLoadResult(
                    endpoint = endpoint,
                    options = cachedOptions,
                    source = OrganizationSource.CACHE,
                    fetchedAt = cached.fetchedAt,
                    error = "${error.message ?: "组织架构服务不可用"}；正在使用上次成功加载的数据",
                )
            } else {
                OrganizationLoadResult(endpoint = endpoint, error = error.message ?: "无法加载组织架构")
            }
        }
    }

    fun save(settings: ReportingSettings): ReportingSettings {
        val normalized = ReportingSupport.normalizeSettings(settings)
        runGitAi(listOf("config", "reporting-profile", "set", "--stdin"), ReportingSupport.toCliJson(normalized))
        return readGitAiSettings()
    }

    private fun readGitAiSettings(): ReportingSettings = ReportingSupport.fromCliJson(
        runGitAi(listOf("config", "reporting-profile")),
    )

    private fun runGitAi(args: List<String>, input: String? = null): String {
        val binary = gitAiService.resolveGitAiBinary()
            ?: throw IllegalStateException("未找到 Git AI CLI，请先安装 Git AI 后重试")
        val process = ProcessBuilder(listOf(binary) + args).redirectErrorStream(false).start()
        val stdout = CompletableFuture.supplyAsync { readCapped(process.inputStream) }
        val stderr = CompletableFuture.supplyAsync { readCapped(process.errorStream) }
        if (input != null) {
            process.outputStream.bufferedWriter(StandardCharsets.UTF_8).use { it.write(input) }
        } else {
            process.outputStream.close()
        }
        if (!process.waitFor(CLI_TIMEOUT_SECONDS, TimeUnit.SECONDS)) {
            process.destroyForcibly()
            throw IllegalStateException("Git AI CLI 响应超时，请确认 CLI 已安装后重试")
        }
        val output = stdout.get()
        val errors = stderr.get()
        if (process.exitValue() != 0) throw IllegalStateException(formatCliError(errors.ifBlank { output }))
        return output.trim()
    }

    private fun readCapped(stream: java.io.InputStream): String {
        stream.use { input ->
            val output = ByteArrayOutputStream()
            val buffer = ByteArray(DEFAULT_BUFFER_SIZE)
            while (true) {
                val count = input.read(buffer)
                if (count < 0) break
                if (output.size() + count > MAX_CLI_OUTPUT_BYTES) {
                    throw IllegalStateException("Git AI CLI 输出过大，已中止本次操作")
                }
                output.write(buffer, 0, count)
            }
            return output.toString(StandardCharsets.UTF_8)
        }
    }

    private fun formatCliError(message: String): String {
        if (Regex("Unknown config key:\\s*reporting[-_]profile").containsMatchIn(message)) {
            return "当前 Git AI CLI 版本不支持数据上报配置，请升级到 1.6.13 或更高版本。"
        }
        return message.ifBlank { "Git AI CLI 执行失败" }
    }

    companion object {
        private const val CLI_TIMEOUT_SECONDS = 10L
        private const val MAX_CLI_OUTPUT_BYTES = 1_000_000
        private const val ORGANIZATION_CACHE_MAX_AGE_MS = 7 * 24 * 60 * 60 * 1000L
    }
}
